Unidirectional cluster links (#18111)
canonical/lxd: doc/explanation/clusters.md

(exp-cluster-links)=
## Cluster links

> Cluster links enable secure communication between separate LXD clusters by pinning the remote cluster's TLS certificate and optionally establishing mutual trust.

Reviewer: I believe that we want to avoid making absolute statements about security. How about reverting back to "authenticated communication," or, if that is no longer appropriate, simply:

```suggestion
Cluster links enable communication between separate LXD clusters by pinning the remote cluster's TLS certificate and optionally establishing mutual trust.
```

Cluster links are the foundation for {ref}`replicators <exp-replicators>`, which use bidirectional links to sync instances across clusters for active-passive disaster recovery. Unidirectional links are suited to scenarios where one cluster needs to read from another without granting reciprocal access, or for anonymous access to a cluster that exposes resources publicly.

### Link types

There are three link types, each suited to different trust and access requirements:

`bidirectional`
: Both clusters authenticate each other using mutual TLS. Each side creates an identity for the other and can initiate requests to the other. This is the default type.

> `unidirectional`

Reviewer: I would suggest that we be consistent with the terminology of the different link types across the sections that follow. How about:

- bidirectional
- unidirectional (authenticated)
- unidirectional unauthenticated

I've placed "authenticated" in parentheses, since it does not appear in the code-formatted text. So, for this entry, how about:

```suggestion
`unidirectional` (authenticated)
```

> : A pins B's certificate and uses a token to activate a pending identity that B created for A. B stores a corresponding link entry and can authenticate incoming requests from A, but holds no address for A and cannot initiate requests to it.

Reviewer: A proposal for revising the "A" and "B" cluster names to more descriptive terms:

```suggestion
: One cluster sends requests and the other cluster receives and authenticates the requests.
To create this type of link, the request recipient creates a pending identity for sender.
The sender pins the receiver's certificate and uses a token to activate the pending identity.
The receiver stores a corresponding link entry and can authenticate incoming requests from the sender, but the receiver does not hold an address for the sender and cannot initiate requests in the other direction.
```

Also, you have a nice description of the use case for `unidirectional-unathenticated` link type. Can you add a similar description to the `unidirectional` link type?

`unidirectional-unauthenticated`

> : Only the initiating cluster (A) stores a link to B. A connects to B without presenting a client certificate, relying solely on certificate pinning for server authentication. B is completely unaware of the link and no identity is created on either side. Use this type when B exposes resources publicly or when you want read-only, anonymous-style access to B.

Reviewer: ```suggestion
: One cluster sends unauthenticated requests to another cluster.
The sender stores a link to the request receiver, but the receiver does not create an identity for the sender.
The sender connects to the receiver without presenting a client certificate, and relies solely on certificate pinning for server authentication.
The receiver remains unaware of the link and does not create an identity for the sender.
Use this type when the receiving cluster exposes resources publicly, or when you want read-only, anonymous-style access to a cluster.
```

> ### How cluster links work

Reviewer: Since this is explanatory material, how about:

```suggestion
### Connection process
```

> All link types rely on TLS certificate pinning: A fetches and pins B's certificate before making any connection. The link type determines what trust is established on B's side.

Reviewer: ```suggestion
All link types rely on TLS certificate pinning: one cluster fetches and pins the other cluster's certificate before making any connection.
The link type determines what trust is established on the other cluster.
```

#### Bidirectional connection process

> 1. **Cluster A** creates a pending cluster link and generates a trust token.
> 1. **Cluster B** uses this token to establish the connection and send its certificate back.

Reviewer: Expanding on the revisions that I've suggested for clusters "A" and "B":

```suggestion
1. One cluster creates a pending cluster link and generates a trust token.
1. The other cluster uses this token to establish the connection and send its certificate back.
```

1. Both clusters validate certificates and activate the bidirectional link.
1. The link becomes active and both clusters can communicate.

> #### Unidirectional connection process

Reviewer: Following up on my comment above, adding "(authenticated)" for consistency:

```suggestion
#### Unidirectional (authenticated) connection process
```

> 1. **Cluster B** issues a pending identity token using `lxc auth identity create cluster-link/<name>`.
> 1. **Cluster A** consumes the token with `lxc cluster link create <name> --token <token> --unidirectional`, pins B's certificate, and calls back to B to activate the pending identity.
> 1. A has an active link to B with no associated identity. B has an active identity for A and a corresponding link row.

Reviewer: Edits of Cluster A/B. Also, since this is an explanation, not a how-to guide, I think you can drop the commands and instead focus on describing how the links work. The user can follow the link below to the relevant how-to guide for the specifics of the commands.

```suggestion
1. The cluster that will receive requests issues a pending identity token.
1. The cluster that will send requests consumes the token, pins the receiver's certificate, and calls back to the receiver to activate the pending identity.
1. The sender has an active link to the receiver with no associated identity. The receiver has an active identity for the sender and a corresponding link row.
```

> #### Unauthenticated unidirectional connection process

Reviewer: Again, for consistency of the types:

```suggestion
#### Unidirectional unauthenticated connection process
```

> 1. **Cluster A** runs `lxc cluster link create <name> --unauthenticated --remote-address <addr>`.
> 1. The CLI fetches B's certificate and displays the fingerprint for the user to confirm.
> 1. A stores the link locally with B's pinned certificate. B is not contacted and remains unaware of the link.

Reviewer: Again, I propose descriptive language to identify the clusters, and also removing the command in favor of a description of what the clusters do.

```suggestion
1. The cluster that will send requests creates an unauthenticated link to the address of the cluster that will receive requests.
1. The CLI fetches the request recipient's certificate and displays the fingerprint for the user to confirm.
1. The sender stores the link locally with the receiver's pinned certificate.
The cluster that receives requests is not contacted and remains unaware of the link.
```

> For more information, see: {ref}`howto-cluster-links-create`.

Reviewer: How about updating this with a more descriptive sentence, to emphasize that the creation process differs for each link type?

```suggestion
Consult our {ref}`guide to creating cluster links <howto-cluster-links-create>` for details about how to create each type of cluster link.
```

(exp-clusters-links-identity)=

> ### Identity management for cluster links

Reviewer: For consistency with "connection process" above, how about simply:

```suggestion
### Identity management
```

> The identities created depend on the link type:

Reviewer: Can you revise this to include a general statement about what an "identity" contains? The fine-grained authorization section of "Remote API authentication", for example, has the following: "An identity is any authenticated party that makes requests to LXD, including TLS clients."

I was thinking a statement along the lines of "Certain types of cluster links create identities in (for?) the linked clusters. An identity is a ... . These identities depend on the link type:"

Reviewer: Or, along the lines of the previous version: "When you create certain types of cluster links, LXD automatically creates an identify for the linked cluster. [These identities are ... ]."

- **Bidirectional**: LXD creates a `Cluster link certificate` identity on each side. The identity can be in one of two states:
  - **Pending**: A trust token has been generated but the link has not been activated yet.
  - **Active**: Both clusters have exchanged certificates and the link is operational.

> - **Unidirectional (authenticated)**: B creates an identity for A and a corresponding link entry (no addresses, so B cannot reach A, but can manage A's access via `lxc cluster link delete`). A stores B's certificate directly without an associated identity.

Reviewer: This point about addresses seems worth pulling out of the parentheses into a separate sentence. I've also proposed some revisions to replace A/B with descriptive language:

```suggestion
- **Unidirectional (authenticated)**: The cluster that will receive requests creates an identity for the cluster that will send requests.
The receiver also creates a corresponding link entry to manage the sender's access.
This link entry does not contain any addresses, so the receiver cannot initiate requests back to the sender.
The cluster that will send requests stores the receiver's certificate directly without creating an associated identity.
```

- **Unidirectional unauthenticated**: No identity is created on either side.

> Identities are managed using {ref}`fine-grained-authorization`.

Reviewer: Since "fine-grained authorization" fits into the sentence so well, let's use custom link text, rather than the cross referenced title:

```suggestion
Identities are managed using {ref}`fine-grained authorization <fine-grained-authorization>`.
```

> #### Security considerations

Reviewer: Is this information specific to "Identity management for cluster links," or more general? If the latter, then perhaps elevate the header level?

```suggestion
### Security considerations
```

- **Identity isolation**: Each cluster link gets its own identity that can be managed independently.
- **Group membership**: Cluster link identities can be assigned to authentication groups for bulk permission management.

> Together, these controls limit the blast radius of a compromised link by enforcing certificate-based trust and least-privilege access. They also make it possible to revoke a single link's access without impacting other cluster-to-cluster trust relationships.

Reviewer: This was in the original, I know, but I believe that we want to avoid language such as "blast" (similarly, "explosive", "kill", "terminate", etc.). Is there an alternative way of describing this? Maybe simply "impact"? Or "potential impact"?

```suggestion
Together, these controls limit the potential impact of a compromised link by enforcing certificate-based trust and least-privilege access.
They also make it possible to revoke a single link's access without impacting other cluster-to-cluster trust relationships.
```

> #### Cluster link deletion

Reviewer: How about "removal"? And, as with the security section, how about elevating the heading level (this doesn't seem nested under "Identity management")?

```suggestion
### Removal
```

Deletion behavior depends on the link type:

> - **Bidirectional**: Run `lxc cluster link delete` on both clusters to fully remove the trust relationship.
> - **Unidirectional (authenticated)**: Deleting on A removes only A's link. B's identity and link row remain until B explicitly deletes its link.
> - **Unidirectional unauthenticated**: Only A has a link. Run `lxc cluster link delete` on A; B is unaffected.

Reviewer: Doesn't this amount to the same thing for both bidirectional and unidirectional (authenticated) cluster links? In essence, it's necessary to delete the link on both clusters, right? How about simply:

```suggestion
- **Bidirectional**: Run `lxc cluster link delete` on both clusters to fully remove the trust relationship.
- **Unidirectional (authenticated)**: Run `lxc cluster link delete` on both clusters to fully remove the link.
- **Unidirectional unauthenticated**: Only the cluster that sends requests has a link.
Run `lxc cluster link delete` on that cluster to remove the link.
```

Reviewer: I had another thought on this part. At the moment, this is closer to step by step instructions for deleting the links. Those steps might be better in the how to guide (run `lxc cluster link delete` on this cluster, then run it again on that cluster). For this part, how about just a short summary, similar to what you've put in the guide? For example, this summary tries to explain the cause and effect behind the different removal process:

```suggestion
The link removal process varies by link type, because the command to delete a cluster link only removes link entries and identities on the local cluster on which you run the command.
In the case of bidirectional and unidirectional (unauthenticated) cluster links, both clusters store information about the cluster link, and you must run the command on both clusters to fully remove the cluster link.
In the case of unidirectional unauthenticated cluster links, only the cluster that sends requests stores information about the cluster link, and you only need to run the command on that cluster to remove the link.
For details on the steps required to remove different types of cluster links, refer to our {ref}`guide to deleting cluster links <howto-cluster-links-delete>`.
```

> #### Cluster link member status

Reviewer: And maybe raise this a level, as well? Since I don't believe that the content is nested under "Identity management"

```suggestion
### Member status
```
